url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/MessagePack.java

```diff
@@ -0,0 +1,837 @@
+
+    /**
+     * Returns serializer that enables serializing objects into
+     * {@link java.io.OutputStream} object.
+     * 
+     * @since 0.6.0
+     * @param out
+     *            output stream
+     * @return stream-based serializer
+     */
+    public Packer createPacker(OutputStream out) {
+        return new MessagePackPacker(this, out);
+    }
+
+    /**
+     * Returns serializer that enables serializing objects into buffer.
+     * 
+     * @since 0.6.0
+     * @return buffer-based serializer
+     */
+    public BufferPacker createBufferPacker() {
+        return new MessagePackBufferPacker(this);
+    }
+
+    /**
+     * Returns serializer that enables serializing objects into buffer.
+     * 
+     * @since 0.6.0
+     * @param bufferSize
+     *            initial size of buffer
+     * @return buffer-based serializer
+     */
+    public BufferPacker createBufferPacker(int bufferSize) {
+        return new MessagePackBufferPacker(this, bufferSize);
+    }
+
+    /**
+     * Returns deserializer that enables deserializing
+     * {@link java.io.InputStream} object.
+     * 
+     * @since 0.6.0
+     * @param in
+     *            input stream
+     * @return stream-based deserializer
+     */
+    public Unpacker createUnpacker(InputStream in) {
+        return new MessagePackUnpacker(this, in);
+    }
+
+    /**
+     * Returns empty deserializer that enables deserializing buffer.
+     * 
+     * @since 0.6.0
+     * @return buffer-based deserializer
+     */
+    public BufferUnpacker createBufferUnpacker() {
+        return new MessagePackBufferUnpacker(this);
+    }
+
+    /**
+     * Returns deserializer that enables deserializing buffer.
+     * 
+     * @since 0.6.0
+     * @param bytes
+     *            input byte array
+     * @return buffer-based deserializer
+     */
+    public BufferUnpacker createBufferUnpacker(byte[] bytes) {
+        return createBufferUnpacker().wrap(bytes);
+    }
+
+    /**
+     * Returns deserializer that enables deserializing buffer.
+     * 
+     * @since 0.6.0
+     * @param bytes
+     * @param off
+     * @param len
+     * @return buffer-based deserializer
+     */
+    public BufferUnpacker createBufferUnpacker(byte[] bytes, int off, int len) {
+        return createBufferUnpacker().wrap(bytes, off, len);
+    }
+
+    /**
+     * Returns deserializer that enables deserializing buffer.
+     * 
+     * @since 0.6.0
+     * @param buffer
+     *            input {@link java.nio.ByteBuffer} object
+     * @return buffer-based deserializer
+     */
+    public BufferUnpacker createBufferUnpacker(ByteBuffer buffer) {
+        return createBufferUnpacker().wrap(buffer);
+    }
+
+    /**
+     * Serializes specified object.
+     * 
+     * @since 0.6.0
+     * @param v
+     *            serialized object
+     * @return output byte array
+     * @throws IOException
+     */
+    public <T> byte[] write(T v) throws IOException {
+        BufferPacker pk = createBufferPacker();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *pk*. Execution paths that do not contain closure statements were detected.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).